Add IndexingParametersConfiguration for Search

heaths opened the pull request:
Resolves #12374

pakrym commented on sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParameters.cs:
```diff
@@ -1,20 +1,230 @@
+            /// <summary>
+            /// Initializes the <see cref="IndexingParametersConfiguration"/> on the parent <see cref="IndexingParameters"/> if null.
+            /// </summary>
+            private void EnsureInitialized() =>
```
Why do you need to lazy-initialize the _indexingParametersConfiguration? I would think we should lazy initialize the configuration dictionary instead.

heaths replied:
Which configuration dictionary? If you mean `IndexingPArameters.Configuration`, that was an `IDictionary<string, object>` before and never null. Sure, we could lazy-init it but I don't see much point. The only reason I'm doing this here is to avoid a potential concurrency issue when initializing the now-model on the parent. The deserializer allows it to be null, which wasn't true for the old `configuration`. If I had some way to provide some hint to the serializing, that would help; however, that's not so straight forward since the `IndexingParametersConfiguration` as a dictionary is only reflecting the `AdditionalProperties` - not itself. That's what the aggregating dictionary is for.

pakrym commented on sdk/search/Azure.Search.Documents/src/Utilities/AggregateDictionary.cs:
```diff
@@ -0,0 +1,311 @@
+    /// </summary>
+    /// <typeparam name="TKey">The type of key.</typeparam>
+    /// <typeparam name="TValue">The type of value.</typeparam>
+    internal abstract class AggregateDictionary<TKey, TValue> : IDictionary<TKey, TValue>
```
I suspect implementation might be simpler if we make it less generic and more specific to the IndexingParametersConfiguration case. It would also be more efficient because you won't have to allocate 2 extra dictionaries.

heaths replied:
I'm not allocating 2 extra dictionaries. I'm allocating an array to reference to existing dictionaries. I'll also be redoing this slightly to avoid `IEnumerator<T>` allocations, reducing GC pressure and keeping with stack memory.

I originally started this making it specific to the case at hand, but if we're going to keep doing JSON objects with well-known properties and additional properties this way, I figured having an abstraction that handles these cases ready should be easier to consume. It also helped separate the logic and testability a bit, but could be convinced to just collapse them into a single definition.

heaths replied:
This is gone now.

heaths commented on sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersConfiguration.cs:
```diff
@@ -0,0 +1,277 @@
+        private IDictionary<string, object> _aggregate;
+
+        /// <summary> Represents the parsing mode for indexing from an Azure blob data source. </summary>
+        public BlobIndexerParsingMode? ParsingMode
```
Thanks for explaining what you meant offline, but that won't work easily and is why I turned this into a backing dictionary in the first place: `Count`, `Keys`, and `Values` all need to grow and shrink as values are set/unset. Lots of other operations would have to check if values are default. This results in far more code for every method in an `IDictionary<TKey, TValue>` and `ICollection<T>` for `Keys` and `Values` than what you see here, meaning more to maintain and more change of errors.

heaths commented on sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParameters.cs:
```diff
@@ -1,20 +1,230 @@
+        [CodeGenMember("Configuration")]
+        public IndexingParametersConfiguration IndexingParametersConfiguration
+        {
+            get => _indexingParametersConfiguration;
```
You can't use a `ref` on a property, which I use elsewhere.

heaths commented on sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParameters.cs:
```diff
@@ -1,20 +1,230 @@
+        /// Base class for read-only keys and values collections for <see cref="ConfigurationAdapter"/>.
+        /// </summary>
+        /// <typeparam name="T">The type of element.</typeparam>
+        private abstract class ReadOnlyCollection<T> : ICollection<T>
```
It's a nested private class so the name doesn't matter. There's no chance of collision. .NET has tons of "name collisions" like this that are also nested.

bleroy replied:
I'm not worried about a collision, but about confusion. The name does matter in that reading the client code for that private class, it's easy to assume it's the one from the BCL.

heaths replied:
While I don't think there should be much confusion around a nested class, I changed it anyway while I almost completely rewrote all this based on some great ideas from @pakrym.

heaths commented on sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersConfiguration.cs:
```diff
@@ -0,0 +1,277 @@
+            set => Set(IndexedFileNameExtensionsKey, value);
+        }
+
+        /// <summary> For Azure blobs, set to false if you want to continue indexing when an unsupported content type is encountered, and you don&apos;t know all the content types (file extensions) in advance. </summary>
```
This is the comment directly from swagger. You might open an issue on Azure/azure-rest-api-specs to improve it, but I couldn't offer any insight to what null means either.

heaths commented on sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersConfiguration.cs:
```diff
@@ -0,0 +1,277 @@
+            set => Set(ExcludedFileNameExtensionsKey, value);
+        }
+
+        /// <summary> Comma-delimited list of filename extensions to select when processing from Azure blob storage.  For example, you could focus indexing on specific application files &quot;.docx, .pptx, .msg&quot; to specifically include those file types. </summary>
```
These are all from swagger.

heaths commented on sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersConfiguration.cs:
```diff
@@ -0,0 +1,277 @@
+        }
+
+        /// <summary> For CSV blobs, specifies the end-of-line single-character delimiter for CSV files where each line starts a new document (for example, &quot;|&quot;). </summary>
+        public string DelimitedTextDelimiter
```
Nothing authored in swagger. As such, we don't send it back (i.e. undefined) so any service defaults apply. Same as other models in Search.

heaths commented on sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersConfiguration.cs:
```diff
@@ -0,0 +1,277 @@
+            set => Set(FirstLineContainsHeadersKey, value);
+        }
+
+        /// <summary> For JSON arrays, given a structured or semi-structured document, you can specify a path to the array using this property. </summary>
```
This is taken directly from swagger. I recommend opening an issue to improve the documentation if this is not sufficient. All languages will have this same text. In fact, I'm making a change where I won't even be copying what is there, but allowing the generator to write it all directly from swagger so no change to override the message.

bleroy commented on sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersConfiguration.cs:
```diff
@@ -0,0 +1,277 @@
+            set => Set(AllowSkillsetToReadFileDataKey, value);
+        }
+
+        /// <summary> Determines algorithm for text extraction from PDF files in Azure blob storage. </summary>
```
I can pack that one into a swagger issue as well.

heaths commented on sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersConfiguration.cs:
```diff
@@ -0,0 +1,277 @@
+        }
+
+        /// <summary> Increases the timeout beyond the 5-minute default for Azure SQL database data sources, specified in the format &quot;hh:mm:ss&quot;. </summary>
+        public string QueryTimeout
```
From swagger

bleroy replied:
Can you even have a TimeSpan in the spec or is this a .NET-ism? Since this is not generated code, I thought we could be more idiomatic than the spec can be.

heaths replied:
It's actually from generated code, but I might be able to declare it as a TimeSpan, or at least have a private field and make this a TimeSpan. I agree that would be better in this case. For string variants, while I appreciate what the old extension methods did, validating the format goes against our guidelines since it can change. But in this case, I think it makes sense.

Please be sure to unresolve any comments you want me to follow-up on. This was still resolved, but I just happened to notice it. I resolve comments as I deal with them so that unresolved make a sort of "todo list".

heaths commented on sdk/search/Azure.Search.Documents/src/Utilities/AggregateDictionary.cs:
```diff
@@ -0,0 +1,311 @@
+                    throw new ArgumentNullException(nameof(key));
+                }
+
+                foreach (IDictionary<TKey, TValue> dictionary in _dictionaries)
+                {
+                    if (dictionary.TryGetValue(key, out TValue value))
+                    {
+                        return value;
+                    }
+                }
```
Those collections are references and have to grow or shrink as the dictionary does. They don't hold onto anything. This is to make sure the existing `IDictionary<string, object> Configuration { get; }` continues to work exactly as it does today.

heaths commented on sdk/search/Azure.Search.Documents/tests/Models/IndexingParametersConfigurationTests.cs:
```diff
@@ -0,0 +1,241 @@
+            {
+                Configuration =
+                {
+                    ["parsingMode"] = "json",
```
Yes. They get saved to the dictionary differently and I have to test both ways. Customers could do both.

heaths commented on sdk/search/Azure.Search.Documents/src/Utilities/AggregateDictionary.cs:
```diff
@@ -0,0 +1,311 @@
+        }
+
+        /// <inheritdoc />
+        public bool Remove(KeyValuePair<TKey, TValue> item)
```
That wouldn't be consistent with how the old dictionary works. `Remove(string key)` doesn't check the value. This variant of `Remove` has to, and does in the stock implementations in the .NET BCL. As mentioned elsewhere, all this is to make sure the existing `Configuration` property works exactly as it did before. We want to avoid any API or behavioral breaking changes.

heaths commented on sdk/search/Azure.Search.Documents/src/Utilities/AggregateDictionary.cs:
```diff
@@ -0,0 +1,311 @@
+            // Assume only 2 elements.
+            const int capacity = 2;
+
+            List<ICollection<TKey>> keys = new List<ICollection<TKey>>(capacity);
+            List<ICollection<TValue>> values = new List<ICollection<TValue>>(capacity);
+
+            foreach (IDictionary<TKey, TValue> dictionary in _dictionaries)
+            {
+                keys.Add(dictionary.Keys);
+                values.Add(dictionary.Values);
+            }
+
+            Keys = new ReadOnlyAggregateCollection<TKey>(keys);
+            Values = new ReadOnlyAggregateCollection<TValue>(values);
```
`ToList` makes a snapshot. `Keys` and `Values` has to grow and shrink. If someone gets these references and holds onto them, a snapshot will not suffice.

The aggregate already does a sufficient job to avoid duplicates. It may not be completely impossible, but given that the JSON source object can't have duplicates (invalid JSON), the two source dictionaries we use shouldn't either. And if customers try to add a duplicate, the aggregate detects this and reacts accordingly. Anything else would be exceptional, thus requiring debugging and appropriate resolution.

bleroy replied:
OK, I see `KeysAndValuesGrow` is testing that.

heaths commented on sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParameters.cs:
```diff
@@ -1,20 +1,230 @@
+        {
+            private readonly IndexingParameters _parent;
+
+            public ConfigurationAdapter(IndexingParameters parent)
```
Other constructors are already defined elsewhere (generated source goes to a different folder). These are partial, so with the other constructors it's already not default creatable.

heaths commented on sdk/search/Azure.Search.Documents/src/Utilities/AggregateDictionary.cs:
```diff
@@ -0,0 +1,311 @@
+            public void Clear() => throw new NotSupportedException();
+
+            /// <inheritdoc />
+            public bool Contains(T item)
```
Docs say it terminates as soon as a match is found, so that could work. That said, most of this is probably getting rewritten anyway.

heaths replied:
Done

heaths committed:
Adding IndexingParametersConfiguration for Search

heaths committed:
Redefine aggregate using getters/setters

Much cleaner design and relies more on the generated code than the previous iteration.

heaths committed:
Update documentation

pakrym commented on sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParameters.cs:
```diff
@@ -1,20 +1,416 @@
+                        return 0;
+                    }
+
+                    // Have to count ourselves since Enumerable.Count() will call into this property causing a StackOverflowException.
```
Whoops :)

heaths replied:
The concept was there. I had to make a few changes anyway, but your founding idea with the dictionary of delegates was great! I got the dictionary I wanted to handle most of the abstraction, and without maintaining separate dictionaries as before to aggregate multiple other dictionaries. This is much cleaner. 😉

pakrym reviewed (approved):
Nice!

pakrym commented on sdk/search/Azure.Search.Documents/src/Indexes/Models/IndexingParametersConfiguration.cs:
```diff
@@ -0,0 +1,87 @@
+        /// <summary>
+        /// Increases the timeout beyond the 5-minute default for Azure SQL database data sources.
+        /// </summary>
+        public TimeSpan? QueryTimeout
```
Did you hit https://github.com/Azure/autorest.csharp/issues/824?

heaths replied:
No, this worked fine. I can't quite tell from the bug, but maybe I avoided it because it was just a string before (in the format "hh:mm:ss"). I regenerated after adding `[CodeGenMember("QueryTimeout")]` to a string field and everything went fine.

heaths committed:
Add more tests for IndexingParameters
